File: apps/dsa-web/src/api/portfolio.ts
import apiClient from './index';
import { toCamelCase } from './utils';
import type { TaskAccepted } from '../types/analysis';
import type {
  ImageImportCommitResponse,
  PositionImageCommitRequest,
  PositionImageParseResponse,
  PortfolioAccountItem,
  PortfolioAccountCreateRequest,
  PortfolioAccountListResponse,
  PortfolioCashLedgerCreateRequest,
  PortfolioCashLedgerListResponse,
  PortfolioCorporateActionCreateRequest,
  PortfolioCorporateActionListResponse,
  PortfolioCostMethod,
  PortfolioDeleteResponse,
  PortfolioEventCreatedResponse,
  PortfolioFxRefreshResponse,
  PortfolioImageDraftUpdateRequest,
  PortfolioImageTaskAccepted,
  PortfolioImageTaskCurrentResponse,
  PortfolioImageTaskSnapshot,
  PortfolioImportBrokerListResponse,
  PortfolioImportCommitResponse,
  PortfolioImportParseResponse,
  PortfolioPositionAnalysisRequest,
  PortfolioRiskResponse,
  PortfolioSnapshotResponse,
  PortfolioTradeCreateRequest,
  PortfolioTradeListResponse,
  TradeImageCommitRequest,
  TradeImageParseResponse,
} from '../types/portfolio';

type SnapshotQuery = {
  accountId?: number;
  asOf?: string;
  costMethod?: PortfolioCostMethod;
  includeRealtime?: boolean;
};

type FxRefreshQuery = {
  accountId?: number;
  asOf?: string;
};

type EventQuery = {
  accountId?: number;
  dateFrom?: string;
  dateTo?: string;
  page?: number;
  pageSize?: number;
};

type TradeListQuery = EventQuery & {
  symbol?: string;
  side?: 'buy' | 'sell';
};

type CashListQuery = EventQuery & {
  direction?: 'in' | 'out';
};

type CorporateListQuery = EventQuery & {
  symbol?: string;
  actionType?: 'cash_dividend' | 'split_adjustment';
};

function buildSnapshotParams(query: SnapshotQuery): Record<string, string | number> {
  const params: Record<string, string | number> = {};
  if (query.accountId != null) {
    params.account_id = query.accountId;
  }
  if (query.asOf) {
    params.as_of = query.asOf;
  }
  if (query.costMethod) {
    params.cost_method = query.costMethod;
  }
  if (query.includeRealtime !== undefined) {
    params.include_realtime = query.includeRealtime ? 'true' : 'false';
  }
  return params;
}

function buildFxRefreshParams(query: FxRefreshQuery): Record<string, string | number> {
  const params: Record<string, string | number> = {};
  if (query.accountId != null) {
    params.account_id = query.accountId;
  }
  if (query.asOf) {
    params.as_of = query.asOf;
  }
  return params;
}

function buildEventParams(query: EventQuery): Record<string, string | number> {
  const params: Record<string, string | number> = {};
  if (query.accountId != null) {
    params.account_id = query.accountId;
  }
  if (query.dateFrom) {
    params.date_from = query.dateFrom;
  }
  if (query.dateTo) {
    params.date_to = query.dateTo;
  }
  if (query.page != null) {
    params.page = query.page;
  }
  if (query.pageSize != null) {
    params.page_size = query.pageSize;
  }
  return params;
}

function buildImageImportFormData(
  accountId: number,
  dateField: 'snapshot_date' | 'default_trade_date',
  dateValue: string,
  files: File[],
): FormData {
  const formData = new FormData();
  formData.append('account_id', String(accountId));
  formData.append(dateField, dateValue);
  files.forEach((file) => formData.append('files', file));
  return formData;
}

export const portfolioApi = {
  async getAccounts(includeInactive = false): Promise<PortfolioAccountListResponse> {
    const response = await apiClient.get<Record<string, unknown>>('/api/v1/portfolio/accounts', {
      params: { include_inactive: includeInactive },
    });
    return toCamelCase<PortfolioAccountListResponse>(response.data);
  },

  async createAccount(payload: PortfolioAccountCreateRequest): Promise<PortfolioAccountItem> {
    const response = await apiClient.post<Record<string, unknown>>('/api/v1/portfolio/accounts', {
      name: payload.name,
      broker: payload.broker,
      market: payload.market,
      base_currency: payload.baseCurrency,
      owner_id: payload.ownerId,
    });
    return toCamelCase<PortfolioAccountItem>(response.data);
  },

  async deleteAccount(accountId: number): Promise<PortfolioDeleteResponse> {
    const response = await apiClient.delete<Record<string, unknown>>(`/api/v1/portfolio/accounts/${accountId}`);
    return toCamelCase<PortfolioDeleteResponse>(response.data);
  },

  async getSnapshot(query: SnapshotQuery = {}): Promise<PortfolioSnapshotResponse> {
    const response = await apiClient.get<Record<string, unknown>>('/api/v1/portfolio/snapshot', {
      params: buildSnapshotParams(query),
    });
    return toCamelCase<PortfolioSnapshotResponse>(response.data);
  },

  async analyzePosition(symbol: string, payload: PortfolioPositionAnalysisRequest = {}): Promise<TaskAccepted> {
    const response = await apiClient.post<Record<string, unknown>>(
      `/api/v1/portfolio/positions/${encodeURIComponent(symbol)}/analysis`,
      {
        account_id: payload.accountId,
        analysis_phase: payload.analysisPhase ?? 'auto',
        force: payload.force ?? false,
      },
    );
    return toCamelCase<TaskAccepted>(response.data);
  },

  async getRisk(query: SnapshotQuery = {}): Promise<PortfolioRiskResponse> {
    const response = await apiClient.get<Record<string, unknown>>('/api/v1/portfolio/risk', {
      params: buildSnapshotParams(query),
    });
    return toCamelCase<PortfolioRiskResponse>(response.data);
  },

  async refreshFx(query: FxRefreshQuery = {}): Promise<PortfolioFxRefreshResponse> {
    const response = await apiClient.post<Record<string, unknown>>('/api/v1/portfolio/fx/refresh', undefined, {
      params: buildFxRefreshParams(query),
    });
    return toCamelCase<PortfolioFxRefreshResponse>(response.data);
  },

  async createTrade(payload: PortfolioTradeCreateRequest): Promise<PortfolioEventCreatedResponse> {
    const response = await apiClient.post<Record<string, unknown>>('/api/v1/portfolio/trades', {
      account_id: payload.accountId,
      symbol: payload.symbol,
      trade_date: payload.tradeDate,
      trade_time: payload.tradeTime,
      side: payload.side,
      quantity: payload.quantity,
      price: payload.price,
      fee: payload.fee ?? 0,
      tax: payload.tax ?? 0,
      market: payload.market,
      currency: payload.currency,
      trade_uid: payload.tradeUid,
      note: payload.note,
    });
    return toCamelCase<PortfolioEventCreatedResponse>(response.data);
  },

  async deleteTrade(tradeId: number): Promise<PortfolioDeleteResponse> {
    const response = await apiClient.delete<Record<string, unknown>>(`/api/v1/portfolio/trades/${tradeId}`);
    return toCamelCase<PortfolioDeleteResponse>(response.data);
  },

  async createCashLedger(payload: PortfolioCashLedgerCreateRequest): Promise<PortfolioEventCreatedResponse> {
    const response = await apiClient.post<Record<string, unknown>>('/api/v1/portfolio/cash-ledger', {
      account_id: payload.accountId,
      event_date: payload.eventDate,
      direction: payload.direction,
      amount: payload.amount,
      currency: payload.currency,
      note: payload.note,
    });
    return toCamelCase<PortfolioEventCreatedResponse>(response.data);
  },

  async deleteCashLedger(entryId: number): Promise<PortfolioDeleteResponse> {
    const response = await apiClient.delete<Record<string, unknown>>(`/api/v1/portfolio/cash-ledger/${entryId}`);
    return toCamelCase<PortfolioDeleteResponse>(response.data);
  },

  async createCorporateAction(payload: PortfolioCorporateActionCreateRequest): Promise<PortfolioEventCreatedResponse> {
    const response = await apiClient.post<Record<string, unknown>>('/api/v1/portfolio/corporate-actions', {
      account_id: payload.accountId,
      symbol: payload.symbol,
      effective_date: payload.effectiveDate,
      action_type: payload.actionType,
      market: payload.market,
      currency: payload.currency,
      cash_dividend_per_share: payload.cashDividendPerShare,
      split_ratio: payload.splitRatio,
      note: payload.note,
    });
    return toCamelCase<PortfolioEventCreatedResponse>(response.data);
  },

  async deleteCorporateAction(actionId: number): Promise<PortfolioDeleteResponse> {
    const response = await apiClient.delete<Record<string, unknown>>(`/api/v1/portfolio/corporate-actions/${actionId}`);
    return toCamelCase<PortfolioDeleteResponse>(response.data);
  },

  async listTrades(query: TradeListQuery = {}): Promise<PortfolioTradeListResponse> {
    const params = buildEventParams(query);
    if (query.symbol) {
      params.symbol = query.symbol;
    }
    if (query.side) {
      params.side = query.side;
    }
    const response = await apiClient.get<Record<string, unknown>>('/api/v1/portfolio/trades', { params });
    return toCamelCase<PortfolioTradeListResponse>(response.data);
  },

  async listCashLedger(query: CashListQuery = {}): Promise<PortfolioCashLedgerListResponse> {
    const params = buildEventParams(query);
    if (query.direction) {
      params.direction = query.direction;
    }
    const response = await apiClient.get<Record<string, unknown>>('/api/v1/portfolio/cash-ledger', { params });
    return toCamelCase<PortfolioCashLedgerListResponse>(response.data);
  },

  async listCorporateActions(query: CorporateListQuery = {}): Promise<PortfolioCorporateActionListResponse> {
    const params = buildEventParams(query);
    if (query.symbol) {
      params.symbol = query.symbol;
    }
    if (query.actionType) {
      params.action_type = query.actionType;
    }
    const response = await apiClient.get<Record<string, unknown>>('/api/v1/portfolio/corporate-actions', { params });
    return toCamelCase<PortfolioCorporateActionListResponse>(response.data);
  },

  async listImportBrokers(): Promise<PortfolioImportBrokerListResponse> {
    const response = await apiClient.get<Record<string, unknown>>('/api/v1/portfolio/imports/csv/brokers');
    return toCamelCase<PortfolioImportBrokerListResponse>(response.data);
  },

  async parseCsvImport(broker: string, file: File): Promise<PortfolioImportParseResponse> {
    const formData = new FormData();
    formData.append('broker', broker);
    formData.append('file', file);
    const response = await apiClient.post<Record<string, unknown>>('/api/v1/portfolio/imports/csv/parse', formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
    });
    return toCamelCase<PortfolioImportParseResponse>(response.data);
  },

  async commitCsvImport(
    accountId: number,
    broker: string,
    file: File,
    dryRun = false,
  ): Promise<PortfolioImportCommitResponse> {
    const formData = new FormData();
    formData.append('account_id', String(accountId));
    formData.append('broker', broker);
    formData.append('dry_run', dryRun ? 'true' : 'false');
    formData.append('file', file);
    const response = await apiClient.post<Record<string, unknown>>('/api/v1/portfolio/imports/csv/commit', formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
    });
    return toCamelCase<PortfolioImportCommitResponse>(response.data);
  },

  async parsePositionImages(
    accountId: number,
    snapshotDate: string,
    files: File[],
  ): Promise<PositionImageParseResponse> {
    const formData = buildImageImportFormData(accountId, 'snapshot_date', snapshotDate, files);
    const response = await apiClient.post<Record<string, unknown>>(
      '/api/v1/portfolio/imports/images/positions/parse',
      formData,
      { headers: { 'Content-Type': 'multipart/form-data' } },
    );
    return toCamelCase<PositionImageParseResponse>(response.data);
  },

  async submitPositionImageTask(
    accountId: number,
    snapshotDate: string,
    files: File[],
  ): Promise<PortfolioImageTaskAccepted> {
    const formData = buildImageImportFormData(accountId, 'snapshot_date', snapshotDate, files);
    const response = await apiClient.post<Record<string, unknown>>(
      '/api/v1/portfolio/imports/images/positions/tasks',
      formData,
      { headers: { 'Content-Type': 'multipart/form-data' } },
    );
    return toCamelCase<PortfolioImageTaskAccepted>(response.data);
  },

  async commitPositionImages(request: PositionImageCommitRequest): Promise<ImageImportCommitResponse> {
    const response = await apiClient.post<Record<string, unknown>>(
      '/api/v1/portfolio/imports/images/positions/commit',
      {
        batch_id: request.batchId,
        account_id: request.accountId,
        snapshot_date: request.snapshotDate,
        task_id: request.taskId,
        expected_revision: request.expectedRevision,
        positions: request.positions.map((position) => ({
          symbol: position.symbol,
          name: position.name,
          quantity: position.quantity,
          avg_cost: position.avgCost,
        })),
      },
    );
    return toCamelCase<ImageImportCommitResponse>(response.data);
  },

  async parseTradeImages(
    accountId: number,
    defaultTradeDate: string,
    files: File[],
  ): Promise<TradeImageParseResponse> {
    const formData = buildImageImportFormData(accountId, 'default_trade_date', defaultTradeDate, files);
    const response = await apiClient.post<Record<string, unknown>>(
      '/api/v1/portfolio/imports/images/trades/parse',
      formData,
      { headers: { 'Content-Type': 'multipart/form-data' } },
    );
    return toCamelCase<TradeImageParseResponse>(response.data);
  },

  async submitTradeImageTask(
    accountId: number,
    defaultTradeDate: string,
    files: File[],
  ): Promise<PortfolioImageTaskAccepted> {
    const formData = buildImageImportFormData(accountId, 'default_trade_date', defaultTradeDate, files);
    const response = await apiClient.post<Record<string, unknown>>(
      '/api/v1/portfolio/imports/images/trades/tasks',
      formData,
      { headers: { 'Content-Type': 'multipart/form-data' } },
    );
    return toCamelCase<PortfolioImageTaskAccepted>(response.data);
  },

  async commitTradeImages(request: TradeImageCommitRequest): Promise<ImageImportCommitResponse> {
    const response = await apiClient.post<Record<string, unknown>>(
      '/api/v1/portfolio/imports/images/trades/commit',
      {
        batch_id: request.batchId,
        account_id: request.accountId,
        task_id: request.taskId,
        expected_revision: request.expectedRevision,
        trades: request.trades.map((trade) => ({
          trade_date: trade.tradeDate,
          trade_time: trade.tradeTime,
          symbol: trade.symbol,
          name: trade.name,
          side: trade.side,
          quantity: trade.quantity,
          price: trade.price,
          fee: trade.fee,
          tax: trade.tax,
          trade_uid: trade.tradeUid,
          occurrence_index: trade.occurrenceIndex,
        })),
      },
    );
    return toCamelCase<ImageImportCommitResponse>(response.data);
  },

  async getCurrentImageTask(): Promise<PortfolioImageTaskCurrentResponse> {
    const response = await apiClient.get<Record<string, unknown>>('/api/v1/portfolio/imports/images/tasks/current');
    return toCamelCase<PortfolioImageTaskCurrentResponse>(response.data);
  },

  async getImageTask(taskId: string): Promise<PortfolioImageTaskSnapshot> {
    const response = await apiClient.get<Record<string, unknown>>(
      `/api/v1/portfolio/imports/images/tasks/${encodeURIComponent(taskId)}`,
    );
    return toCamelCase<PortfolioImageTaskSnapshot>(response.data);
  },

  async updateImageTaskDraft(
    taskId: string,
    request: PortfolioImageDraftUpdateRequest,
  ): Promise<PortfolioImageTaskSnapshot> {
    const response = await apiClient.patch<Record<string, unknown>>(
      `/api/v1/portfolio/imports/images/tasks/${encodeURIComponent(taskId)}/draft`,
      {
        expected_revision: request.expectedRevision,
        files: request.files,
        positions: request.positions?.map((position) => ({
          source_refs: position.sourceRefs.map((ref) => ({ file_index: ref.fileIndex, row_index: ref.rowIndex })),
          symbol: position.symbol,
          name: position.name,
          quantity: position.quantity,
          avg_cost: position.avgCost,
          current_price: position.currentPrice,
          market_value: position.marketValue,
          available_quantity: position.availableQuantity,
          weight_pct: position.weightPct,
          profit_loss: position.profitLoss,
          confidence: position.confidence,
          status: position.status,
          issues: position.issues,
        })),
        trades: request.trades?.map((trade) => ({
          source_refs: trade.sourceRefs.map((ref) => ({ file_index: ref.fileIndex, row_index: ref.rowIndex })),
          trade_date: trade.tradeDate,
          trade_time: trade.tradeTime,
          symbol: trade.symbol,
          name: trade.name,
          side: trade.side,
          quantity: trade.quantity,
          price: trade.price,
          fee: trade.fee,
          tax: trade.tax,
          trade_uid: trade.tradeUid,
          confidence: trade.confidence,
          occurrence_index: trade.occurrenceIndex,
          fingerprint: trade.fingerprint,
          dedup_hash: trade.dedupHash,
          status: trade.status,
          issues: trade.issues,
        })),
      },
    );
    return toCamelCase<PortfolioImageTaskSnapshot>(response.data);
  },

  async cancelImageTask(taskId: string): Promise<PortfolioImageTaskSnapshot> {
    const response = await apiClient.post<Record<string, unknown>>(
      `/api/v1/portfolio/imports/images/tasks/${encodeURIComponent(taskId)}/cancel`,
    );
    return toCamelCase<PortfolioImageTaskSnapshot>(response.data);
  },

  async discardImageTask(taskId: string): Promise<void> {
    await apiClient.delete(`/api/v1/portfolio/imports/images/tasks/${encodeURIComponent(taskId)}`);
  },
};

export function getExistingPortfolioImageTaskId(error: unknown): string | null {
  const response = (error as { response?: { data?: Record<string, unknown> } } | null)?.response;
  const data = response?.data;
  if (!data) return null;
  if (typeof data.existing_task_id === 'string') return data.existing_task_id;
  const detail = data.detail;
  if (detail && typeof detail === 'object' && typeof (detail as Record<string, unknown>).existing_task_id === 'string') {
    return (detail as Record<string, unknown>).existing_task_id as string;
  }
  return null;
}

File: api/v1/schemas/portfolio.py
# -*- coding: utf-8 -*-
"""Portfolio API schemas."""

from __future__ import annotations

from datetime import date, time
from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field


class PortfolioAccountCreateRequest(BaseModel):
    name: str = Field(..., min_length=1, max_length=64)
    broker: Optional[str] = Field(None, max_length=64)
    market: Literal["cn", "hk", "us", "jp", "kr", "tw"] = "cn"
    base_currency: str = Field("CNY", min_length=3, max_length=8)
    owner_id: Optional[str] = Field(None, max_length=64)


class PortfolioAccountUpdateRequest(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=64)
    broker: Optional[str] = Field(None, max_length=64)
    market: Optional[Literal["cn", "hk", "us", "jp", "kr", "tw"]] = None
    base_currency: Optional[str] = Field(None, min_length=3, max_length=8)
    owner_id: Optional[str] = Field(None, max_length=64)
    is_active: Optional[bool] = None


class PortfolioAccountItem(BaseModel):
    id: int
    owner_id: Optional[str] = None
    name: str
    broker: Optional[str] = None
    market: str
    base_currency: str
    is_active: bool
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class PortfolioAccountListResponse(BaseModel):
    accounts: List[PortfolioAccountItem] = Field(default_factory=list)


class PortfolioTradeCreateRequest(BaseModel):
    account_id: int
    symbol: str = Field(..., min_length=1, max_length=16)
    trade_date: date
    trade_time: Optional[time] = None
    side: Literal["buy", "sell"]
    quantity: float = Field(..., gt=0)
    price: float = Field(..., gt=0)
    fee: float = Field(0.0, ge=0)
    tax: float = Field(0.0, ge=0)
    market: Optional[Literal["cn", "hk", "us", "jp", "kr", "tw"]] = None
    currency: Optional[str] = Field(None, min_length=3, max_length=8)
    trade_uid: Optional[str] = Field(None, max_length=128)
    note: Optional[str] = Field(None, max_length=255)


class PortfolioCashLedgerCreateRequest(BaseModel):
    account_id: int
    event_date: date
    direction: Literal["in", "out"]
    amount: float = Field(..., gt=0)
    currency: Optional[str] = Field(None, min_length=3, max_length=8)
    note: Optional[str] = Field(None, max_length=255)


class PortfolioCorporateActionCreateRequest(BaseModel):
    account_id: int
    symbol: str = Field(..., min_length=1, max_length=16)
    effective_date: date
    action_type: Literal["cash_dividend", "split_adjustment"]
    market: Optional[Literal["cn", "hk", "us", "jp", "kr", "tw"]] = None
    currency: Optional[str] = Field(None, min_length=3, max_length=8)
    cash_dividend_per_share: Optional[float] = Field(None, ge=0)
    split_ratio: Optional[float] = Field(None, gt=0)
    note: Optional[str] = Field(None, max_length=255)


class PortfolioEventCreatedResponse(BaseModel):
    id: int


class PortfolioDeleteResponse(BaseModel):
    deleted: int


class PortfolioTradeListItem(BaseModel):
    id: int
    account_id: int
    trade_uid: Optional[str] = None
    symbol: str
    market: str
    currency: str
    trade_date: str
    trade_time: Optional[str] = None
    side: str
    quantity: float
    price: float
    fee: float
    tax: float
    note: Optional[str] = None
    created_at: Optional[str] = None


class PortfolioTradeListResponse(BaseModel):
    items: List[PortfolioTradeListItem] = Field(default_factory=list)
    total: int
    page: int
    page_size: int


class PortfolioCashLedgerListItem(BaseModel):
    id: int
    account_id: int
    event_date: str
    direction: str
    amount: float
    currency: str
    note: Optional[str] = None
    created_at: Optional[str] = None


class PortfolioCashLedgerListResponse(BaseModel):
    items: List[PortfolioCashLedgerListItem] = Field(default_factory=list)
    total: int
    page: int
    page_size: int


class PortfolioCorporateActionListItem(BaseModel):
    id: int
    account_id: int
    symbol: str
    market: str
    currency: str
    effective_date: str
    action_type: str
    cash_dividend_per_share: Optional[float] = None
    split_ratio: Optional[float] = None
    note: Optional[str] = None
    created_at: Optional[str] = None


class PortfolioCorporateActionListResponse(BaseModel):
    items: List[PortfolioCorporateActionListItem] = Field(default_factory=list)
    total: int
    page: int
    page_size: int


class PortfolioPositionItem(BaseModel):
    symbol: str
    market: str
    currency: str
    quantity: float
    avg_cost: float
    total_cost: float
    last_price: float
    market_value_base: float
    unrealized_pnl_base: float
    unrealized_pnl_pct: Optional[float] = None
    valuation_currency: str
    price_source: str = "unknown"
    price_provider: Optional[str] = None
    price_date: Optional[str] = None
    price_stale: bool = False
    price_available: bool = True
    data_quality: str = "ok"
    limitations: List[str] = Field(default_factory=list)


class PortfolioPositionAnalysisRequest(BaseModel):
    account_id: Optional[int] = Field(None, description="Optional account id; required when a symbol is held in multiple accounts")
    analysis_phase: Literal["auto", "premarket", "intraday", "postmarket"] = "auto"
    force: bool = Field(False, description="Force refresh analysis inputs without bypassing duplicate in-flight tasks")


class PortfolioAccountSnapshot(BaseModel):
    account_id: int
    account_name: str
    owner_id: Optional[str] = None
    broker: Optional[str] = None
    market: str
    base_currency: str
    as_of: str
    cost_method: str
    total_cash: float
    total_market_value: float
    total_equity: float
    realized_pnl: float
    unrealized_pnl: float
    fee_total: float
    tax_total: float
    fx_stale: bool
    data_quality: str = "ok"
    limitations: List[str] = Field(default_factory=list)
    positions: List[PortfolioPositionItem] = Field(default_factory=list)


class PortfolioSnapshotResponse(BaseModel):
    as_of: str
    cost_method: str
    currency: str
    account_count: int
    total_cash: float
    total_market_value: float
    total_equity: float
    realized_pnl: float
    unrealized_pnl: float
    fee_total: float
    tax_total: float
    fx_stale: bool
    data_quality: str = "ok"
    limitations: List[str] = Field(default_factory=list)
    accounts: List[PortfolioAccountSnapshot] = Field(default_factory=list)


class PortfolioImportTradeItem(BaseModel):
    trade_date: str
    symbol: str
    side: Literal["buy", "sell"]
    quantity: float
    price: float
    fee: float
    tax: float
    trade_uid: Optional[str] = None
    dedup_hash: str
    currency: Optional[str] = None


class PortfolioImportParseResponse(BaseModel):
    broker: str
    record_count: int
    skipped_count: int
    error_count: int
    records: List[PortfolioImportTradeItem] = Field(default_factory=list)
    errors: List[str] = Field(default_factory=list)


class PortfolioImportCommitResponse(BaseModel):
    account_id: int
    record_count: int
    inserted_count: int
    duplicate_count: int
    failed_count: int
    dry_run: bool
    errors: List[str] = Field(default_factory=list)


class PortfolioImportBrokerItem(BaseModel):
    broker: str
    aliases: List[str] = Field(default_factory=list)
    display_name: Optional[str] = None


class PortfolioImportBrokerListResponse(BaseModel):
    brokers: List[PortfolioImportBrokerItem] = Field(default_factory=list)


class PortfolioImageFileResult(BaseModel):
    index: int
    filename: Optional[str] = None
    status: Literal["success", "failed"]
    record_count: int = 0
    error: Optional[str] = None


PortfolioImageTaskStatus = Literal[
    "pending",
    "processing",
    "cancel_requested",
    "cancelled",
    "review_required",
    "committing",
    "failed",
]


class PortfolioImageTaskFileItem(BaseModel):
    index: int
    filename: Optional[str] = None
    status: Literal["pending", "processing", "success", "failed", "cancelled"]
    record_count: int = 0
    error: Optional[str] = None
    removed: bool = False


class PortfolioImageTaskAccepted(BaseModel):
    task_id: str
    trace_id: str
    status: PortfolioImageTaskStatus
    mode: Literal["positions", "trades"]
    account_id: int
    account_name: str
    message: str


class PortfolioImageTaskSnapshot(BaseModel):
    task_id: str
    trace_id: str
    mode: Literal["positions", "trades"]
    account_id: int
    account_name: str
    status: PortfolioImageTaskStatus
    message: str
    error_code: Optional[str] = None
    snapshot_date: Optional[str] = None
    default_trade_date: Optional[str] = None
    created_at: str
    started_at: Optional[str] = None
    finished_at: Optional[str] = None
    files: List[PortfolioImageTaskFileItem] = Field(default_factory=list)
    current_file_index: Optional[int] = None
    total_files: int = 0
    current_attempt: Optional[int] = None
    max_attempts: int = 2
    success_count: int = 0
    failure_count: int = 0
    batch_id: Optional[str] = None
    draft_revision: Optional[int] = None
    draft: Optional[Dict[str, Any]] = None


class PortfolioImageTaskCurrentResponse(BaseModel):
    task: Optional[PortfolioImageTaskSnapshot] = None


class PortfolioImageDraftFileUpdate(BaseModel):
    index: int
    removed: bool = False


class PortfolioImageSourceRef(BaseModel):
    file_index: int
    row_index: int


class PortfolioPositionImageItem(BaseModel):
    source_refs: List[PortfolioImageSourceRef] = Field(default_factory=list)
    symbol: str
    name: str
    quantity: Optional[float] = None
    avg_cost: Optional[float] = None
    current_price: Optional[float] = None
    market_value: Optional[float] = None
    available_quantity: Optional[float] = None
    weight_pct: Optional[float] = None
    profit_loss: Optional[float] = None
    confidence: Literal["high", "medium", "low"] = "low"
    status: Literal["ready", "conflict", "error"]
    issues: List[str] = Field(default_factory=list)


class PortfolioPositionImageParseResponse(BaseModel):
    batch_id: str
    account_id: int
    snapshot_date: str
    files: List[PortfolioImageFileResult] = Field(default_factory=list)
    summary: Dict[str, Optional[float]] = Field(default_factory=dict)
    positions: List[PortfolioPositionImageItem] = Field(default_factory=list)


class PortfolioTradeImageItem(BaseModel):
    source_refs: List[PortfolioImageSourceRef] = Field(default_factory=list)
    trade_date: str
    trade_time: Optional[str] = None
    symbol: str
    name: str
    side: str
    quantity: Optional[float] = None
    price: Optional[float] = None
    fee: float = 0
    tax: float = 0
    trade_uid: Optional[str] = None
    confidence: Literal["high", "medium", "low"] = "low"
    occurrence_index: int = 1
    fingerprint: str = ""
    dedup_hash: Optional[str] = None
    status: Literal["ready", "conflict", "error"]
    issues: List[str] = Field(default_factory=list)


class PortfolioTradeImageParseResponse(BaseModel):
    batch_id: str
    account_id: int
    default_trade_date: str
    files: List[PortfolioImageFileResult] = Field(default_factory=list)
    trades: List[PortfolioTradeImageItem] = Field(default_factory=list)


class PortfolioImageDraftUpdateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    expected_revision: int = Field(..., ge=1)
    files: List[PortfolioImageDraftFileUpdate] = Field(default_factory=list)
    positions: Optional[List[PortfolioPositionImageItem]] = None
    trades: Optional[List[PortfolioTradeImageItem]] = None


class PortfolioPositionImageCommitItem(BaseModel):
    model_config = ConfigDict(extra="forbid")

    symbol: str = Field(..., pattern=r"^\d{6}$")
    name: str = Field(..., min_length=1, max_length=64)
    quantity: float = Field(..., gt=0)
    avg_cost: float = Field(..., gt=0)


class PortfolioPositionImageCommitRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    batch_id: str = Field(..., min_length=1, max_length=64)
    account_id: int
    snapshot_date: date
    positions: List[PortfolioPositionImageCommitItem] = Field(..., min_length=1)
    task_id: Optional[str] = Field(None, min_length=1, max_length=64)
    expected_revision: Optional[int] = Field(None, ge=1)


class PortfolioTradeImageCommitItem(BaseModel):
    model_config = ConfigDict(extra="forbid")

    trade_date: date
    trade_time: Optional[time] = None
    symbol: str = Field(..., pattern=r"^\d{6}$")
    name: Optional[str] = Field(None, max_length=64)
    side: Literal["buy", "sell"]
    quantity: float = Field(..., gt=0)
    price: float = Field(..., gt=0)
    fee: float = Field(0, ge=0)
    tax: float = Field(0, ge=0)
    trade_uid: Optional[str] = Field(None, max_length=128)
    occurrence_index: int = Field(1, ge=1)


class PortfolioTradeImageCommitRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    batch_id: str = Field(..., min_length=1, max_length=64)
    account_id: int
    trades: List[PortfolioTradeImageCommitItem] = Field(..., min_length=1)
    task_id: Optional[str] = Field(None, min_length=1, max_length=64)
    expected_revision: Optional[int] = Field(None, ge=1)


class PortfolioImageImportCommitResponse(BaseModel):
    record_count: int
    inserted_count: int
    duplicate_count: int
    failed_count: int
    errors: List[str] = Field(default_factory=list)


class PortfolioFxRefreshResponse(BaseModel):
    as_of: str
    account_count: int
    refresh_enabled: bool
    disabled_reason: Optional[str] = None
    pair_count: int
    updated_count: int
    stale_count: int
    error_count: int


class PortfolioDecisionSignalRiskItem(BaseModel):
    account_id: Optional[int] = None
    symbol: str
    market: str
    signal: Dict[str, Any] = Field(default_factory=dict)


class PortfolioDecisionSignalRiskBlock(BaseModel):
    available: bool = True
    total: int = 0
    actions: Dict[str, int] = Field(default_factory=dict)
    items: List[PortfolioDecisionSignalRiskItem] = Field(default_factory=list)


class PortfolioRiskResponse(BaseModel):
    as_of: str
    account_id: Optional[int] = None
    cost_method: str
    currency: str
    thresholds: Dict[str, Any] = Field(default_factory=dict)
    concentration: Dict[str, Any] = Field(default_factory=dict)
    sector_concentration: Dict[str, Any] = Field(default_factory=dict)
    drawdown: Dict[str, Any] = Field(default_factory=dict)
    stop_loss: Dict[str, Any] = Field(default_factory=dict)
    decision_signal_risk: PortfolioDecisionSignalRiskBlock = Field(default_factory=PortfolioDecisionSignalRiskBlock)
